/* $Id$ */
--- src/send.c Wed Aug  3 22:37:59 2005
+++ src/send.c  Mon Aug 15 00:57:15 2005
@@ -65,6 +65,7 @@
 send_format(char *lsendbuf, int bufsize, const char *pattern, va_list args)
 {
   int len;
+  va_list args_copy;

   /*
    * from rfc1459
@@ -78,7 +79,10 @@
    * continuation message lines.  See section 7 for more details about
    * current implementations.
    */
-  len = vsnprintf(lsendbuf, bufsize - 1, pattern, args);
+
+  va_copy(args_copy, args);
+  len = vsnprintf(lsendbuf, bufsize - 1, pattern, args_copy);
+  va_end(args_copy);
   if (len > bufsize - 2)
     len = bufsize - 2;  /* required by some versions of vsnprintf */
